src/Symfony/Component/Clock/DateTime.php

```diff
@@ -0,0 +1,123 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Clock;
+
+/**
+ * An immmutable DateTime with stricter error handling and return types than the native one.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class DateTime extends \DateTimeImmutable
+{
+    /**
+     * @throws \DateMalformedStringException When $datetime is invalid
+     */
+    public function __construct(string $datetime = 'now', \DateTimeZone $timezone = null)
+    {
+        $now = Clock::get()->now();
+
+        if (null !== $timezone) {
+            $now = $now->setTimezone($timezone);
+        }
+
+        if ('now' !== $datetime) {
+            if (!$now instanceof static) {
```

Comment:
Maybe all these occurrences of `static` can be replaced by `self` (including return types) since this class is final.

Reply:
I prefer keeping `static`, it's the way to advertise wither APIs.

Reply:
Whiters should differentiate returning `$this` instead of `self` or static`, which is not yet possible natively.
If the class is final, there are no benefits, and a drawback since the resolution is made at runtime instead of compilation.

Reply:
It's the job of the engine to optimize `static` into `self` for final classes, not ours ;)
About withers, we do have a way to differentiate: `@return $this`.